analytic_engine/src/sst/parquet/hybrid.rs

```diff
@@ -0,0 +1,441 @@
+// Copyright 2022 CeresDB Project Authors. Licensed under Apache-2.0.
+
+use std::{collections::BTreeMap, sync::Arc};
+
+use arrow_deps::arrow::{
+    array::{Array, ArrayData, ArrayRef, ListArray, StringArray, UInt64Array},
+    bitmap::Bitmap,
+    buffer::MutableBuffer,
+    datatypes::Schema as ArrowSchema,
+    record_batch::RecordBatch as ArrowRecordBatch,
+    util::bit_util,
+};
+use common_types::{
+    datum::DatumKind,
+    schema::{ArrowSchemaRef, DataType, Field, Schema},
+};
+use log::debug;
+use snafu::{OptionExt, ResultExt};
+
+use crate::sst::builder::{EncodeRecordBatch, Result, VariableLengthType};
+
+//  hard coded in https://github.com/apache/arrow-rs/blob/20.0.0/arrow/src/array/array_list.rs#L185
+const LIST_ITEM_NAME: &str = "item";
+
+/// ArrayHandle is used to keep different offsets of array, which can be concat
+/// together.
+///
+/// Note:
+/// 1. Array.slice(offset, length) don't work as expected, since the
+/// underlying buffer is still shared without slice.
+/// 2. Array shoule be [fixed-size primitive](https://arrow.apache.org/docs/format/Columnar.html#fixed-size-primitive-layout)
+#[derive(Debug, Clone)]
+struct ArrayHandle {
+    array: ArrayRef,
+    positions: Vec<(usize, usize)>, // (offset ,length)
```

Comment:
Is the `length` unnecessary (because it can be computed by `positions[idx+1] - positions[idx]`)?